src/zeta/aca_zeta_programming.cpp

```diff
@@ -206,39 +209,35 @@ int ACA_Zeta_Programming::create_zeta_config(const alcor::schema::AuxGateway cur
     oam_port_listener_thread->detach();
     ACA_LOG_INFO("Created thread for port %d and it is detached.\n", oam_port);
   } else {
-    if (current_zeta_cfg->zeta_buckets.hashSize !=
-        (uint)current_AuxGateway.destinations().size()) {
-      bucket_not_found = true;
-    } else {
-      for (auto destination : current_AuxGateway.destinations()) {
-        FWD_Info *target_fwd =
-                new FWD_Info(destination.ip_address(), destination.mac_address());
-
-        int *not_used = nullptr;
-
-        if (current_zeta_cfg->zeta_buckets.find(target_fwd, not_used)) {
-          continue;
-        } else {
-          bucket_not_found |= true;
-          break;
-        }
-        new_zeta_buckets.insert(target_fwd, nullptr);
+    for (auto destination : current_AuxGateway.destinations()) {
+      FWD_Info target_fwd(destination.ip_address(), destination.mac_address());
+
+      int *not_used = nullptr;
+
+      if (current_zeta_cfg->zeta_buckets.find(target_fwd, not_used)) {
+        continue;
+      } else {
+        bucket_not_found |= true;
+        break;
       }
+      new_zeta_buckets.insert(target_fwd, nullptr);
     }
 
     // If the buckets have changed, update the buckets and group table rules.
     if (bucket_not_found == true) {
       current_zeta_cfg->zeta_buckets.clear();
       for (auto destination : current_AuxGateway.destinations()) {
-        FWD_Info *target_fwd =
-                new FWD_Info(destination.ip_address(), destination.mac_address());
+        FWD_Info target_fwd =
+                FWD_Info(destination.ip_address(), destination.mac_address());
 
         current_zeta_cfg->zeta_buckets.insert(target_fwd, nullptr);
       }
 
       overall_rc = _update_zeta_group_entry(current_zeta_cfg);
     }
   }
+  _zeta_config_table_mutex.unlock();
```

Comment:
see if you can unlock it sooner, like right after "not found and create":

  if (!_zeta_config_table.find(current_AuxGateway.id(), current_zeta_cfg)) {
    create_entry(current_AuxGateway.id(), oam_port, current_AuxGateway);
    _zeta_config_table.find(current_AuxGateway.id(), current_zeta_cfg);
    oam_port_listener_thread->detach();
    ACA_LOG_INFO("Created thread for port %d and it is detached.\n", oam_port);
  }

I will let you think about it :)